contracts/Staking.sol

```diff
@@ -0,0 +1,488 @@
+    uint8 eventCode = uint8(iter.next().toUint());
+    uint32 resCode;
+    if (eventCode == EVENT_TRANSFER_IN_REWARD) {
+      resCode = _handleTransferInRewardSynPackage(iter);
+    } else if (eventCode == EVENT_TRANSFER_IN_UNDELEGATED) {
+      resCode = _handleTransferInUndelegatedSynPackage(iter);
+    } else {
+      resCode = ERROR_UNKNOWN_PACKAGE_TYPE;
+    }
+
+    if (resCode == CODE_OK) {
+      return new bytes(0);
+    } else {
+      emit failedSynPackage(resCode);
+      return msgBytes;
+    }
+  }
+
+  function handleAckPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract onlyInit override {
+    RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
+    uint8 eventCode = uint8(iter.next().toUint());
+    if (eventCode == EVENT_DELEGATE) {
+      _handleDelegateAckPackage(iter);
+    } else if (eventCode == EVENT_UNDELEGATE) {
+      _handleUndelegateAckPackage(iter);
+    } else if (eventCode == EVENT_REDELEGATE) {
+      _handleRedelegateAckPackage(iter);
+    } else {
+      require(false, "unknown event type");
+    }
+    return;
+  }
+
+  function handleFailAckPackage(uint8, bytes calldata) external onlyCrossChainContract onlyInit override {
+    emit crashResponse();
+    return;
+  }
+
+  /***************************** External functions *****************************/
+  function delegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(amount >= minDelegationChange, "the amount must not be less than minDelegationChange");
+    require(msg.value >= amount.add(oracleRelayerFee), "the msg value should be no less than the sum of stake amount and minimum oracleRelayerFee");
+
+    // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);
+    uint256 _oracleRelayerFee = (msg.value).sub(amount);
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_DELEGATE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_DELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(msg.value);
+
+    delegated[msg.sender] = delegated[msg.sender].add(amount);
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].add(amount);
+
+    emit delegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function undelegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(pendingUndelegated[msg.sender][validator] == 0, "pending undelegation exist");
+    amount = amount != 0 ? amount : delegatedOfValidator[msg.sender][validator];
```

Comment:
Because the msg.sender can get the exact value, the amount should be specified precisely. I think there is no need special case for 0 amount, just forbid it.

Reply:
Ok